Dependencies: migrate from python-Levenshtein to pylev

jayaddison opened the pull request:
This is a tentative suggestion, and doesn't include performance impact analysis.

[`pylev`](https://github.com/toastdriven/pylev) is a pure-python implementation of the Levenshtein distance calculation, and I believe it can be used as a drop-in replacement for [`python-Levenshtein`](https://github.com/miohtama/python-Levenshtein).

The main purpose of the switch to `pylev` for the use case I'm looking at is that it removes a number of container build-time dependencies, including the need to compile C code at build-time.  Switching to `pylev` should (I believe) make `fast-autocomplete` usable in pure-Python container environments (`pygraphviz` and some other optional dependencies may still require C compilation steps).

NB: `pylev` does use a slightly more permissive license (New BSD rather than GPL) which could also factor into vendor preference.

seperman commented:
Hi @jayaddison 
Thanks for making the PR.
How is the performance of pylev compared to the C library? Maybe then the users should choose which option to use by adding it in the setup.py. For example the setup.py will let them choose by doing: `pip install fast-autocomplete[levenshtein]` or `pip install fast-autocomplete[pylev]`.
Levenshtein edit distance algorithm is slow to begin with so using a slower implementation should be optional in my mind.

jayaddison commented:
Thanks @seperman - anecdotally it looks like `pylev` may require 30x (!) CPU time on my hardware; I haven't benchmarked memory usage.  I'll admit that's more overhead than I anticipated.

Your idea about using optional library dependencies sounds good, although I may need to do a little more research to figure out whether `pylev` can be specified as an _optional_ dependency that _replaces_ (essentially overrides) an existing (Levenshtein) dependency.  Default library extras don't [appear](https://stackoverflow.com/questions/60842775/how-to-package-a-python-module-with-extras-as-default) to be possible, but I'll do more digging.

Just for reference, the benchmarking data and runtime scripts are included below:

```python
# pair-generation.py
from random import choice, shuffle
  
with open('/usr/share/dict/words', 'r') as f:
    words = f.readlines()

pairs = []
for a in words:
    a = a.strip()
    b = choice(words).strip()
    pairs.append((a, b))

shuffle(pairs)
for pair in pairs:
    a, b = pair
    print(f'{a}, {b}')
```

```python
# workload-baseline.py
from Levenshtein import distance as levenshtein

with open('pairs.txt', 'r') as f:
    pairs = f.readlines()

for pair in pairs:
    a, b = pair.split(', ')
    levenshtein(a, b)
```

```python
# workload-pylev.py
from pylev import levenshtein

with open('pairs.txt', 'r') as f:
    pairs = f.readlines()

for pair in pairs:
    a, b = pair.split(', ')
    levenshtein(a, b)
```

```bash
$ time python workload-baseline.py 

real	0m0.113s
user	0m0.112s
sys	0m0.000s

$ time python workload-pylev.py 

real	0m3.181s
user	0m3.167s
sys	0m0.012s
```

seperman commented:
Ok thanks for the update! Yeah usually the C libraries should be used instead of Python ones if the C one is available. Also one of these days when I have time, I will add some optimizations so the Levenshtein algorithm is used even less often as it is very expensive in CPU usage.
I don't think memory usage should be a concern for autocomplete usage since those will be small phrases.

jayaddison commented:
@seperman The latest few commits here allow users to specify either `levenshtein` or `pylev` as the distance calculation module at install-time.

This is a breaking change since a straightforward `pip install fast-autocomplete` with these changes would not install either of the dependencies.  It does provide a (hopefully useful) error message under those conditions.

A workaround would be to add a [post-install step](https://stackoverflow.com/questions/53376135/install-by-default-optional-dependencies-in-python-setuptools) to ensure that an implementation is present, or install one if not.

That'd allow the change to be non-breaking, but also adds custom code execution at install-time (not ideal), and could be tricky to unpick in future (especially if migrating to the `wheel` Python package format in future, which as far as I understand does not support post-install scripts).

jayaddison commented:
Let me know what you reckon about the possible approach(es) @seperman - it's possible I'm overthinking the problem; glad for any feedback when you have time.

seperman commented:
Also I have moved this repo from my employer's repo to my personal account after getting the legal approval. Github will automatically redirect to the new address. Just wanted to keep you posted!

jayaddison commented:
Awesome, and thanks for the heads-up :)
